src/vm/codepitchingmanager.cpp

```diff
@@ -0,0 +1,526 @@
+        return FALSE;
+
+    if (IsLCGMethod() || IsVtableMethod() || IsInterface() || IsVirtual())
+        return FALSE;
+
+    if (s_pNotForPitchingMethodsLock != NULL)
+    {
+        SimpleReadLockHolder srlh(s_pNotForPitchingMethodsLock);
+        if (s_pNotForPitchingMethods != NULL)
+        {
+            UPTR key = (UPTR)GetFullHash(this);
+            MethodDesc *pFound = (MethodDesc *)s_pNotForPitchingMethods->LookupValue(key, (LPVOID)this);
+            if (pFound != (MethodDesc *)INVALIDENTRY)
+            {
+                return FALSE;
+            }
+        }
+    }
+    return TRUE;
+}
+
+
+static BOOL IsOwnerOfRWLock(LPVOID lock)
+{
+  // @TODO - SimpleRWLock does not have knowledge of which thread gets the writer
+  // lock, so no way to verify
+  return TRUE;
+}
+
+EXTERN_C bool LookupOrCreateInNotForPitching(MethodDesc* pMD)
+{
+    CONTRACTL
+    {
+        MODE_COOPERATIVE;
+        GC_TRIGGERS;
+        THROWS;
+    }
+    CONTRACTL_END;
+
+    if (pMD != NULL && pMD->IsPitchable())
+    {
+        CreateRWLock(&s_pNotForPitchingMethodsLock);
+
+        UPTR key = (UPTR)GetFullHash(pMD);
+
+        if (s_pNotForPitchingMethods == NULL)
+        {
+            SimpleWriteLockHolder swlh(s_pNotForPitchingMethodsLock);
+            if (s_pNotForPitchingMethods == NULL)
+            {
+                PtrHashMap *pMap = new (SystemDomain::GetGlobalLoaderAllocator()->GetLowFrequencyHeap()) PtrHashMap();
+                LockOwner lock = {s_pNotForPitchingMethodsLock, IsOwnerOfRWLock};
+                pMap->Init(32, NULL, FALSE, &lock);
+                s_pNotForPitchingMethods = pMap;
+            }
+        }
+        else
+        {
+             SimpleReadLockHolder srlh(s_pNotForPitchingMethodsLock);
+             MethodDesc *pFound = (MethodDesc *)s_pNotForPitchingMethods->LookupValue(key, (LPVOID)pMD);
+             if (pFound != (MethodDesc *)INVALIDENTRY)
+                 return TRUE;
+        }
+
+        {
+             SimpleWriteLockHolder swlh(s_pNotForPitchingMethodsLock);
+             s_pNotForPitchingMethods->InsertValue(key, (LPVOID)pMD);
+        }
+    }
+    return FALSE;
+}
+
+static void LookupOrCreateInPitchingCandidate(MethodDesc* pMD, ULONG sizeOfCode)
+{
+    CONTRACTL
+    {
+        MODE_COOPERATIVE;
+        GC_TRIGGERS;
+        THROWS;
+    }
+    CONTRACTL_END;
+
+    if (!pMD->IsPitchable())
+        return;
+
+    PCODE prCode = pMD->GetPreImplementedCode();
+    if (prCode != NULL)
+        return;
+
+    if (!pMD->HasPrecode())
+        return;
+
+    // We lazily allocate the reader/writer lock we synchronize access to the hash with.
+    CreateRWLock(&s_pPitchingCandidateMethodsLock);
+
+    // Now we have a lock we can use to synchronize the remainder of the init.
+
+    UPTR key = (UPTR)GetFullHash(pMD);
+
+    if (s_pPitchingCandidateMethods == NULL)
+    {
+        SimpleWriteLockHolder swlh(s_pPitchingCandidateMethodsLock);
+        if (s_pPitchingCandidateMethods == NULL)
```

Comment:
The lazy initialization of all the maps and locks make the code harder to read, obscuring the real functionality. It doesn't seem to save much. It seems it would be better to just allocate all of them in a single initialization function where you would know that the pitching is active. Did you have any particular reason for using the lazy allocations?